tools/src/main/java/org/apache/kafka/tools/ConnectPluginPath.java

```diff
@@ -0,0 +1,399 @@
+    }
+
+    private static <T> void operation(
+            // fixed
+            Config config,
+            // each location
+            PluginSource source,
+            // each plugin
+            String pluginName,
+            String pluginVersion,
+            List<String> aliases,
+            PluginType pluginType,
+            boolean loadable,
+            boolean hasManifest
+    ) {
+        Path pluginLocation = source.location();
+        if (config.command == Command.LIST) {
+            String firstAlias = aliases.size() > 0 ? aliases.get(0) : "null";
+            String secondAlias = aliases.size() > 1 ? aliases.get(1) : "null";
+            // there should only be one location, otherwise the plugin source column will be ambiguous.
+            config.out.printf(
+                    "%s\t%s\t%s\t%s\t%s\t%s\t%b\t%b%n",
+                    pluginLocation,
+                    pluginName,
+                    firstAlias,
+                    secondAlias,
+                    pluginVersion,
+                    pluginType,
+                    loadable,
+                    hasManifest
+            );
+        }
+    }
+
+    private static class ManifestEntry {
+        private final URI manifestURI;
+        private final String className;
+        private final PluginType type;
+
+        private ManifestEntry(URI manifestURI, String className, PluginType type) {
+            this.manifestURI = manifestURI;
+            this.className = className;
+            this.type = type;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            ManifestEntry that = (ManifestEntry) o;
+            return manifestURI.equals(that.manifestURI) && className.equals(that.className) && type == that.type;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(manifestURI, className, type);
+        }
+    }
+
+    private static Set<ManifestEntry> findManifests(PluginSource source) throws IOException {
+        Set<ManifestEntry> manifests = new HashSet<>();
+        for (PluginType type : PluginType.values()) {
+            if (type == PluginType.UNKNOWN) {
+                continue;
+            }
+            try {
+                Enumeration<URL> resources = source.loader().getResources(MANIFEST_PREFIX + type.typeName());
+                while (resources.hasMoreElements())  {
+                    URL url = resources.nextElement();
+                    for (String className : parse(url)) {
+                        manifests.add(new ManifestEntry(url.toURI(), className, type));
+                    }
+                }
+            } catch (URISyntaxException e) {
+                throw new IOException(e);
+            }
+        }
+        return manifests;
+    }
+
+    // Based on implementation from ServiceLoader.LazyClassPathLookupIterator
```

Comment:
I think I was ripping from [JDK11](https://github.com/AdoptOpenJDK/openjdk-jdk11u/blob/master/src/java.base/share/classes/java/util/ServiceLoader.java#L1109) at the time of writing. I believe the LazyClassPathLookupIterator was introduced during the JDK9 rewrite to have ServiceLoader find module declarations in addition to manifest files.

Reply:
Can we add that info to the comment? If bugs are reported for the implementation that we're adapting, and they also affect our modified version, it may make it easier to connect the dots in the future.